xmake/rules/c++/modules/modules_support/msvc.lua

```diff
@@ -0,0 +1,322 @@
+--!A cross-platform build utility based on Lua
+--
+-- Licensed under the Apache License, Version 2.0 (the "License");
+-- you may not use this file except in compliance with the License.
+-- You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+-- Copyright (C) 2015-present, TBOOX Open Source Group.
+--
+-- @author      ruki
+-- @file        msvc.lua
+--
+
+-- imports
+import("core.tool.compiler")
+import("core.project.project")
+import("core.project.depend")
+import("utils.progress")
+import("private.action.build.object", {alias = "objectbuilder"})
+import("common")
+
+modulesflag = nil
+ifcoutputflag = nil
+ifcsearchdirflag = nil
+interfaceflag = nil
+referenceflag = nil
+headernameflag = nil
+headerunitflag = nil
+exportheaderflag = nil
+stdifcdirflag = nil
+scandependenciesflag = nil
+cxxsourcefileflag = "/TP"
+
+function get_bmi_ext()
+    return ".ifc"
+end
+
+-- load parent target with modules files
+function load_parent(target, opt)
+    for _, dep in ipairs(target:orderdeps()) do
+        cachedir = common.get_cache_dir(dep)
+        target:add("cxxflags", {ifcsearchdirflag, cachedir}, {force = true, expand = false})
+    end
+end
+
+-- check C++20 module support
+function check_module_support(target)
+    local compinst = target:compiler("cxx")
+
+    if compinst:has_flags("/experimental:module", "cxxflags", {flagskey = "cl_experimental_module"}) then
+        modulesflag = "/experimental:module"
+    end
+    assert(modulesflag, "compiler(msvc): does not support c++ module!")
```

Comment:
All flags are checked once for each target, and xmake does not recommend defining global variables directly.

We should use caches and use functions to defer fetching them.

```lua
function get_modulesflag(target)
    local modulesflag = _g.modulesflag
    if modulesflag == nil then
        local compinst = target:compiler("cxx")
        if compinst:has_flags("/experimental:module", "cxxflags", {flagskey = "cl_experimental_module"}) then
            modulesflag = "/experimental:module"
        end
        assert(modulesflag, "compiler(msvc): does not support c++ module!")
        _g.modulesflag = modulesflag or false
    end
    return modulesflag
end
```

Reply:
what is _g ?

Reply:
Each module will have a built-in global object to store all global variables, similar to lua's native _G